auditbeat/module/file_integrity/kprobes/path.go

```diff
@@ -0,0 +1,324 @@
+	ErrC() <-chan error
+	Close() error
+}
+
+type statMatch struct {
+	ino        uint64
+	major      uint32
+	minor      uint32
+	depth      uint32
+	fileName   string
+	isFromMove bool
+	tid        uint32
+	fullPath   string
+}
+
+type pTraverser struct {
+	mtx           sync.RWMutex
+	errC          chan error
+	ctx           context.Context
+	cancelFn      context.CancelFunc
+	e             executor
+	w             inotifyWatcher
+	isRecursive   bool
+	waitQueueChan chan struct{}
+	sMatchTimeout time.Duration
+	statQueue     []statMatch
+}
+
+var lstat = os.Lstat // for testing
+
+func newPathMonitor(ctx context.Context, exec executor, timeOut time.Duration, isRecursive bool) (*pTraverser, error) {
+	mWatcher, err := newInotifyWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	if timeOut == 0 {
+		timeOut = 5 * time.Second
+	}
+
+	mCtx, cancelFn := context.WithCancel(ctx)
+
+	return &pTraverser{
+		mtx:           sync.RWMutex{},
+		ctx:           mCtx,
+		errC:          make(chan error),
+		cancelFn:      cancelFn,
+		e:             exec,
+		w:             mWatcher,
+		isRecursive:   isRecursive,
+		sMatchTimeout: timeOut,
+	}, nil
+}
+
+func (r *pTraverser) Close() error {
+	r.cancelFn()
+	return r.w.Close()
+}
+
+func (r *pTraverser) GetMonitorPath(ino uint64, major uint32, minor uint32, name string) (MonitorPath, bool) {
+	if r.ctx.Err() != nil {
+		return MonitorPath{}, false
+	}
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	if len(r.statQueue) == 0 {
+		return MonitorPath{}, false
+	}
+
+	monitorPath := r.statQueue[0]
+	if monitorPath.ino != ino ||
+		monitorPath.major != major ||
+		monitorPath.minor != minor ||
+		monitorPath.fileName != name {
+		return MonitorPath{}, false
+	}
+
+	r.statQueue = r.statQueue[1:]
+
+	if len(r.statQueue) == 0 && r.waitQueueChan != nil {
+		close(r.waitQueueChan)
+		r.waitQueueChan = nil
+	}
+
+	return MonitorPath{
+		fullPath:   monitorPath.fullPath,
+		depth:      monitorPath.depth,
+		isFromMove: monitorPath.isFromMove,
+		tid:        monitorPath.tid,
+	}, true
+}
+
+func readDirNames(dirName string) ([]string, error) {
+	f, err := os.Open(dirName)
+	if err != nil {
+		return nil, err
+	}
+	names, err := f.Readdirnames(-1)
+	_ = f.Close()
+	if err != nil {
+		return nil, err
+	}
+	sort.Strings(names)
+	return names, nil
+}
+
+func (r *pTraverser) ErrC() <-chan error {
+	return r.errC
+}
+
+func (r *pTraverser) WalkAsync(path string, depth uint32, tid uint32) {
+	if r.ctx.Err() != nil {
+		return
+	}
+
+	go func() {
+		walkErr := r.e.Run(func() error {
+			return r.walk(r.ctx, path, depth, true, tid)
+		})
+
+		if walkErr == nil {
+			return
+		}
+
+		select {
+		case r.errC <- walkErr:
+		case <-r.ctx.Done():
+		}
+
+	}()
+}
+
+func (r *pTraverser) walkRecursive(ctx context.Context, path string, mounts mountPoints, depth uint32, isFromMove bool, tid uint32) error {
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	if r.ctx.Err() != nil {
+		return r.ctx.Err()
+	}
```

Comment:
two contexts can fail here 🤔 , are those referred to two completely different contexts or is one a subset of the other?

Reply:
so `r.ctx` is the context of the `pathTraverser` and this is bound to the monitor context; In other words if this is Done then the kprobes monitor is closed. Now the plain `ctx` is the context of the request to walk this path, which can be a different not directly associated context to this one, so double check it is